### PR TITLE
Implement Search Pagination

### PR DESCRIPTION
**Pull request by aahei**
Implement Search Pagination

## Description

- Create a `SearchPagination` component
- Remove the search result limit on WebSoc fuzzy search (may lower the performance)
In addition:
- Create `stc/helpers/constants.ts` to store `PAGE_SIZE`
- Add the "No results found" prompt.

The current config is:
- fetch all results at a time
- display 10 records on a page

## Screenshots

image

## Final Checks:
- [ ] Verify successful deployment
- [ ] Delete branch

(optional)
- [ ] Write tests
- [ ] Write documentation

### CODE REVIEW

**Review by js0mmer — APPROVED** [review]
LGTM! Great work 🔥

**Comment by js0mmer** [comment]
Also, I'm guessing the issue you opened for performance improvements is meant for a new task/PR. I'm going to go ahead and unlink it from this PR so it doesn't close when this one merges.

**Comment by aahei** [comment]
Just an inaccurate test: 
Before improvement, it takes around 4000-5000 ms to load the course results with an empty query (~5900 results).
After improvement, it takes 2000-3000 ms.

**Comment by aahei** [comment]
Improvement: load only the first 5 pages initially, but when the user jumps to the 3rd page or after, load all pages.

**Comment by aahei** [comment]
Peterportal

> Improvement: load only the first 5 pages initially, but when the user jumps to the 3rd page or after, load all pages.

This is a temporary solution. Currently, PPAPI does not support range search, see https://github.com/icssc/peterportal-client/issues/310. We will change it to the actual range search upon the PPAPI update release.

File: site/src/component/SearchPagination/SearchPagination.tsx [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+import { PAGE_SIZE } from "src/helpers/constants";
+import { useAppDispatch, useAppSelector } from "src/store/hooks";
+import { setPageNumber } from "src/store/slices/searchSlice";
+import { SearchIndex } from "src/types/types";
+
+interface SearchPaginationProps {
+  index: SearchIndex;
+}
+
+const SearchPagination: FC<SearchPaginationProps> = ({ index }) => {
```
Comment (ethanwong16) [inline site/src/component/SearchPagination/SearchPagination.tsx:12]: Can you add some quick comments on this module explaining what it does and what some of the key functions do?

File: site/src/helpers/constants.ts [hunk 2]
```diff
@@ -0,0 +1,2 @@
+// SearchPage
+export const PAGE_SIZE = 10;
```
Comment (ethanwong16) [inline site/src/helpers/constants.ts:2]: I would suggest renaming this to something like `NUM_RESULTS_PER_PAGE` , but I'll leave this up to you :)

**Review by ethanwong16 — CHANGES_REQUESTED** [review]
LGTM!  Just requested some minor changes :)) thanks!

**Comment by aahei** [comment]
Next todo: merge to Master

**Comment by aahei** [comment]
Next todo: test, also test #366

**Review by js0mmer — COMMENTED** [review]
Optimization seems to have broken for going past page 3. Gets stuck in an infinite loop. Dark mode theming needed for pagination component.

**Comment by aahei** [comment]
> Optimization seems to have broken for going past page 3. Gets stuck in an infinite loop. Dark mode theming needed for pagination component.

Thank you for noting this. It didn't happen when I tested it, so it might be some side effect when merging with the new code since this functionality was written a long time ago. I will try to investigate this on this weekend.

**Comment by js0mmer** [comment]
> > Optimization seems to have broken for going past page 3. Gets stuck in an infinite loop. Dark mode theming needed for pagination component.
> 
> Thank you for noting this. It didn't happen when I tested it, so it might be some side effect when merging with the new code since this functionality was written a long time ago. I will try to investigate this on this weekend.

Honestly, I think we can leave out the first 5 pages limiter. I've tested locally and performance seems to be fine with this current version of fuzzy search. I think we should also lower the search delay from 500ms to 300ms so response times are a bit faster.

**Comment by js0mmer** [comment]
> > > Optimization seems to have broken for going past page 3. Gets stuck in an infinite loop. Dark mode theming needed for pagination component.
> > 
> > 
> > Thank you for noting this. It didn't happen when I tested it, so it might be some side effect when merging with the new code since this functionality was written a long time ago. I will try to investigate this on this weekend.
> 
> Honestly, I think we can leave out the first 5 pages limiter. I've tested locally and performance seems to be fine with this current version of fuzzy search. I think we should also lower the search delay from 500ms to 300ms so response times are a bit faster.

Nevermind, seems like it is necessary given that changing between catalogue/roadmap tabs slows down when searching the additional results.

**Comment by js0mmer** [comment]
Found a solution: moving the states for `hasFullResults` and `lastQuery` into the redux slice seems to prevent these infinite loops. Will implement along with theming then this should be good to go.

**Pull request merged**